### DISCUSSION TITLE
Removing Items from Member Browser

### DISCUSSION THREAD

**Discussion by JonFP** [codefori/vscode-ibmi#316] [Q&A]
I have added a file to the member browser but no longer need it there.  How do I get rid of it? There is no right click option or any other way to do it that I can find.

**Comment by worksofliam** [comment]
There is a right-click option on the source file, not the library.

![image](https://user-images.githubusercontent.com/3708366/134378219-ed7e9e39-7353-4364-8a6b-1600d6ff19c4.png)

**Comment by JonFP** [comment]
Methinks I have discovered a bug then - on this particular item there is no such option - but there are on other source files.  I _think_ it may have happened because I specified a file name instead of a library and it accepted it.  It opens to the next level (i.e. the member) and there is a "Remove ..." option on that source member!

I won't try deleting it yet in case there is anything else you need me to look at first.

**Comment by worksofliam** [comment]
@JonFP Would you mind sharing the `sourceFileList` for your connection directly from the json settings in vscode?

thanks!

**Comment by JonFP** [comment]
Be delighted to - but so far I can't find it.  Where is it stored?


Jon Paris

***@***.***

**Comment by worksofliam** [comment]
Press F1 and search `settings json`. It's the middle result:

![image](https://user-images.githubusercontent.com/3708366/134390277-96fe0f57-39c0-4397-b6f2-2a51edebfd57.png)

Then search for `sourceFileList` in that JSON.

**Reply by JonFP** [reply]
OK - this is the relevant portion of the json.
"sourceFileList": [
            "QSYSINC/H",
            "PARTNER400/ITJTIPSRC",
            "PARTNER400/QRPGLESRC",
            "PARTNER400/TECHSOURCE",
            "LIBHTTP/QRPGLESRC",
            "YAJL/QRPGLESRC",
            "PARTNER400/ZOOMSTUFF",
            "PARTNER400/ITJTIPSRC",
            "PARTNER400/WEBCLIENTS",
            "PARTNER400/QCBLLESRC",
            "QRPGLESRC/MIPLAY.RPGLE",  <<<< ===== Error
            "PARTNER400/ITJTIPSRC"
        ],
    
        
    I've highlighted the erroneous entry.    
    
    I've deleted that entry and also the duplicate for ITJTIPSRC. It doesn't seem to notice if you add dups by mistake.

**Comment by priceaj** [comment]
I dont think this is a bug as this works for me (that is the deleting part seems to function correctly), although it thinks QRPGLESRC is a library and I have the option to "remove source file from view" on what it thinks is the source file (MIPLAY.RPGLE), once I take that option it removes the MIPLAY.RPGLE (and with it QRPGLESRC).

I do think that this shouldn't have been allowed to be added through the UI though!

@worksofliam a couple of things regarding this:

1. Do you think it's worth adding some validation into the "Add Source file" option? 

I belive these are the rules for library and object names:
"In the name of a single object, each part (the object name and the library qualifier name) can have a maximum of 10 characters."
"The first character must be alphabetic or one of the special characters, $, @, or #. The remaining characters can be alphanumeric, a period, an underscore, or one of special characters, $, @, or #."

In regex thats `^[A-Za-z\$\@\#][\w\.\$\@\#]{0,9}\/[A-Za-z\$\@\#][\w\.\$\@\#]{0,9}$`

Although the name above would only be caught because "MIPLAY.RPGLE" is longer than 10 characters (the . is valid) 

Additionally should we validate the library and source file actually exist before adding them to the view!?

2. Since "Add source file" in the Member Browser doesnt actually create a source file (source file creation is done through object browser), should we rename to "Add source file to view" to make it less ambiguous?
